syntax = "proto3";

package scissy.pb;

message Void
{
}

enum Status
{
  kSucceed           = 0;
  kFailed            = 1;
  kInvalidSession    = 2;
  kNotImplemented    = 3;
  kInsufficientRight = 4;
  kNotFound          = 5;
  kInternalError     = 6;
  kForbidden         = 7;
}

message StatusMsg
{
  Status status = 1;
  string msg    = 2;
}

message ServiceInfo
{
  bytes  clone_user          = 1;
  bytes  clone_host          = 2;
  bytes  version             = 3;
  uint64 start_time          = 4;
  bool   is_register_enabled = 5;
}

enum Role
{
  kNone   = 0;
  kReader = 1;
  kWriter = 2;
  kOwner  = 3;
}

enum SshKeyType
{
  kSshKeyRsa = 0;
  kSshKeyDsa = 1;
}

message UserAuth
{
  bytes user     = 1;
  bytes password = 2;
}

message Session
{
  bytes  user    = 1;
  int64  user_id = 2;
  bytes  email   = 3;
  Status status  = 4;
  string msg     = 5;
  Role   role    = 6;
}

message UserCreate
{
  bytes user     = 1;
  bytes password = 2;
  bytes email    = 3;
}

message UserPasswordChange
{
  bytes user         = 1;
  bytes old_password = 2;
  bytes new_password = 3;
}

message UserDelete
{
  bytes user = 1;
}

message SshKey
{
  SshKeyType type   = 1;
  bytes      key    = 2;
  string     desc   = 3;
  uint64     key_id = 4;
}

message UserSshKey
{
  SshKey        key  = 1;
}

message UserSshKeySelector
{
  bytes         user = 3; // target
}

message UserSshKeys
{
  repeated SshKey keys   = 1;
  Status status = 2;
  string msg    = 3;
}

message GroupCreate
{
  bytes grp          = 2;
  bytes desc         = 3;
}

message GroupDelete
{
  int64         grp_id = 2;
  bytes         grp    = 3;
}

message GroupAddUser
{
  bytes grp          = 2;
  int64 grp_id       = 3;
  bytes user         = 4;
  int64 user_id      = 5;
  Role  role         = 6;
}

message GroupRemoveUser
{
  bytes grp          = 2;
  int64 grp_id       = 3;
  bytes user         = 4;
  int64 user_id      = 5;
}

message GroupSelector
{
  bytes  grp    = 1; // the name of the group
  bytes  user   = 2; // the group which contains this user
  uint32 limit  = 3;
  uint32 offset = 4;
  int64  grp_id = 5;
}

message GroupMemberSelector
{
  bytes  grp     = 1;
  int64  grp_id  = 2;
  bytes  user    = 3;
  int64  user_id = 4;
  Role   role    = 5;
  uint32 limit   = 6;
  uint32 offset  = 7;
}

message GroupInfo
{
  bytes  grp    = 1;
  uint32 size   = 2;
  int64  grp_id = 3;
  string desc   = 4;
  Status status = 5;
  string msg    = 6;
}

message GroupList
{
  repeated GroupInfo grps   = 1;
  Status    status = 2;
  string    msg    = 3;
}

message Member
{
  Status status  = 1;
  bytes  user    = 2;
  int64  user_id = 3;
  Role   role    = 4;
  bytes  grp     = 5;
  int64  grp_id  = 6;
}

message MemberList
{
  Status status = 1;
  string msg    = 2;
  repeated Member groups = 3;
  repeated Member users  = 4;
}

message RepoCreate
{
  bytes         name      = 2;
  bytes         desc      = 3;
  bool          is_public = 4;
}

message RepoUpdate
{
  int64         repo_id   = 2;
  bytes         repo      = 3;
  bytes         name      = 4;
  bytes         desc      = 5;
  bool          is_public = 6;
}

message RepoDelete
{
  int64         repo_id = 2;
  bytes         repo    = 3;
}

message RepoAddUser
{
  int64         repo_id = 2;
  bytes         repo    = 3;
  int64         user_id = 4;
  bytes         user    = 5;
  Role          role    = 6;
}

message RepoRemoveUser
{
  int64         repo_id = 2;
  bytes         repo    = 3;
  int64         user_id = 4;
  bytes         user    = 5;
}

message RepoAddGroup
{
  int64         repo_id = 2;
  bytes         repo    = 3;
  int64         grp_id  = 4;
  bytes         grp     = 5;
  Role          role    = 6;
}

message RepoRemoveGroup
{
  int64         repo_id = 2;
  bytes         repo    = 3;
  int64         grp_id  = 4;
  bytes         grp     = 5;
}

message RepoName
{
  bytes name = 1;
}

message RepoInfo
{
  Status status = 1;
  string msg    = 2;

  int64  id               = 3;
  bytes  name             = 4;
  bytes  desc             = 5;
  bool   is_public        = 6;
  uint64 last_commit_time = 7;
}

message RepoMembers
{
  Status status = 1;
  string msg    = 2;

  repeated int64  users_id  = 3;
  repeated bytes  users     = 4;
  repeated int64  groups_id = 5;
  repeated bytes  groups    = 6;
}

message RepoList
{
  repeated RepoInfo repos            = 1;
  Status   status           = 2;
  string   msg              = 3;
}

message RepoSelector
{
  uint32        limit     = 1;
  uint32        offset    = 2;
  int64         repo_id   = 3;
  bytes         repo      = 4;
  bytes         commit_id = 6;
}

message CommitSelector
{
  bytes         revision = 2;
  int64         repo_id  = 3;
  bytes         repo     = 4;
}

message LogSelector
{
  bytes         revision = 2;
  int64         repo_id  = 3;
  bytes         repo     = 4;
  uint32        limit    = 5;
  uint32        offset   = 6;
}

message TreeSelector
{
  bytes         revision  = 2;
  int64         repo_id   = 3;
  bytes         repo      = 4;
  bytes         directory = 5;
}

message BlobSelector
{
  bytes         revision  = 2;
  int64         repo_id   = 3;
  bytes         repo      = 4;
  bytes         path      = 5;
}

message DiffSelector
{
  int64         repo_id      = 2;
  bytes         repo         = 3;
  bytes         revision_old = 4;
  bytes         revision_new = 5;
  bytes         path         = 6;
}

message AccessUserRepo
{
  int64 repo_id   = 1;
  int64 user_id   = 2;
}

message ShellControl
{
  int64 user_id   = 1;
  bytes repo_name = 2;
}

message ShellStatus
{
  Status status    = 1;
  bytes  repo_path = 2;
  Role   role      = 3;
  bool   is_public = 4;
  bool   is_admin  = 5;
}

enum GitFileMode
{
  kGitFilemodeNew            = 0000000;
  kGitFilemodeTree           = 0040000;
  kGitFilemodeBlob           = 0100644;
  kGitFilemodeBlobExecutable = 0100755;
  kGitFilemodeLink           = 0120000;
  kGitFilemodeCommit         = 0160000;
}

enum GitOtype
{
  kGitObjDummy    = 0;
  kGitObjCommit   = 1;
  kGitObjTree     = 2;
  kGitObjBlob     = 3;
  kGitObjTag      = 4;
  kGitObjOfsDelta = 6;
  kGitObjRefDelta = 7;
  kGitObjAny      = -2;
  kGitObjBad      = -1;
}

message GitBranch
{
  bytes name = 1;
}

message GitBranches
{
  Status    status   = 1;
  bytes     msg      = 2;
  repeated GitBranch branches = 3;
}

message GitSignature
{
  bytes  name        = 1;
  bytes  email       = 2;
  uint64 time        = 3;
}

message GitCommit
{
  Status       status      = 1;
  bytes        id          = 2;
  repeated bytes        parents_id  = 3;
  bytes        msg         = 4;
  uint64       time        = 5;
  GitSignature commiter    = 6;
  GitSignature author      = 7;
  bytes        tree_id     = 8;
}

message GitLog
{
  Status    status  = 1;
  bytes     msg     = 2;
  repeated GitCommit commits = 3;
}

message GitTreeEntry
{
  bytes       name = 1;
  bytes       id   = 2;
  GitOtype    type = 3;
  GitFileMode mode = 4;
}

message GitTree
{
  Status status        = 1;
  bytes  msg           = 2;
  repeated GitTreeEntry entries = 3;
}

message GitBlob
{
  Status status       = 1;
  bytes  msg          = 2;
  bool   is_binary    = 3;
  bytes  data         = 4;
  bytes  content_type = 5;
}

message GitPatch
{
  Status status = 1;
  bytes  msg    = 2;
  bytes  data   = 3;
}

message GitTag
{
  bytes  name     = 1;
  bytes  revision = 2;
}

message GitTags
{
  Status status = 1;
  bytes  msg    = 2;
  repeated GitTag tags   = 3;
}

service Service
{
  /* service */
  rpc serviceInfo(Void) returns (ServiceInfo);

  /* session */
  rpc userAuth(UserAuth) returns (Session);
  rpc userGetSession(Void) returns (Session);

  /* user management */
  rpc userCreate(UserCreate) returns (StatusMsg);
  // rpc userDelete(UserDelete) returns (StatusMsg);
  rpc userAddSshKey(UserSshKey) returns (StatusMsg);
  rpc userRemoveSshKey(UserSshKey) returns (StatusMsg);
  rpc userGetSshKeys(UserSshKeySelector) returns (UserSshKeys);
  rpc userSetPassword(UserPasswordChange) returns (StatusMsg);

  /* group management */
  rpc groupCreate(GroupCreate) returns (GroupInfo);
  rpc groupDelete(GroupDelete) returns (StatusMsg);
  rpc groupAddUser(GroupAddUser) returns (StatusMsg);
  rpc groupRemoveUser(GroupRemoveUser) returns (StatusMsg);

  /* groups view */
  rpc groupGetInfo(GroupSelector) returns (GroupInfo);
  rpc groupsList(GroupSelector) returns (GroupList);
  rpc groupListMembers(GroupMemberSelector) returns (MemberList);

  /* repo management */
  rpc repoCreate(RepoCreate) returns (RepoInfo);
  rpc repoUpdate(RepoUpdate) returns (StatusMsg);
  rpc repoDelete(RepoDelete) returns (StatusMsg);
  rpc repoAddUser(RepoAddUser) returns (StatusMsg);
  rpc repoRemoveUser(RepoRemoveUser) returns (StatusMsg);
  rpc repoAddGroup(RepoAddGroup) returns (StatusMsg);
  rpc repoRemoveGroup(RepoRemoveGroup) returns (StatusMsg);

  /* repo view */
  rpc reposList(RepoSelector) returns (RepoList);
  rpc repoGetInfo(RepoSelector) returns (RepoInfo);
  rpc repoListMembers(RepoSelector) returns (MemberList);
  rpc repoListBranches(RepoSelector) returns (GitBranches);
  rpc repoGetCommit(CommitSelector) returns (GitCommit);
  rpc repoGetLog(LogSelector) returns (GitLog);
  rpc repoGetTree(TreeSelector) returns (GitTree);
  rpc repoGetBlob(BlobSelector) returns (GitBlob);
  rpc repoGetPatch(DiffSelector) returns (GitPatch);
  rpc repoGetTags(RepoSelector) returns (GitTags);

  /* ssh shell */
  rpc shellControl(ShellControl) returns (ShellStatus);
}
